add description for styleguidist vuese and vue-design-system (#40)
dobromir-hristov/vuecommunity: docs/guide/ecosystem/documentation.md

- [Docsify](https://docsify.js.org/#/) - native javascript alternative to Docute, with a bit more features, plugins and themes.
- [Jekyll](https://jekyllrb.com/)
- [Hexo](https://hexo.io/)
- [GitBook](https://www.gitbook.com/)
- [Docusaurus](https://github.com/facebook/docusaurus)
- [Docz](https://www.docz.site/)

## Component workbenches

A workbench is a development environment for UI components. It allows developers to setup interactive examples of how a UI component can be used. Simply put, you can setup a live demo of a Vue component by passing different props to it, creating a working sandbox.

This is very powerful when working in big teams with lots of custom UI components. A real rendered component with usage example can be easier to understand than just a text example.

### Storybook

The most popular workbench is [storybook](https://storybook.js.org/).

In storybook, demo components (stories) are written in JavaScript or TypeScript. It makes for easy adoption by front-end developers.

Each component you write story for will have its own page on the final built website. You can then publish this website as a documentation for your components, internally or publicly.

> Using HMR, development of new components in isolation is a breeze.

Author: Yep I agree, I will be a bit more "straight to the point".

There are lots of plugins available ranging from accessibility to site layout.

Used by Angular, React and many other big players, it is probably the safest option.

Since designers usually don't like to wreck JavaScript, storybook is the best choice for developers to have authority on how components are showcased, without intervention from designers.

#### Pros

- Big and welcoming community

> - Supports every feature no exception

Author: agreed

> - Supports every feature no exception

Author: I removed the list of feature from Styleguidist

- Lots of [plugins](#essential-plugins) from visual testing, to integration with sketch symbols
- Easy adoption by developers (Examples written in JavaScript)

#### Cons

> - Bare bones without plugins (no prop detection, no code samples, no styling)

Author: fixed on my local

- Input and output format not meant for long descriptions

> - Templates are plain text so no syntax highlighting

Author: Indeed full sentences are better

#### Essential Plugins

- [@storybook/addon-a11y](https://github.com/storybooks/storybook/tree/next/addons/a11y) checks accessibility properties
- [@storybook/addon-knobs](https://github.com/storybooks/storybook) add controls to change usage of your component
- [@storybook/addon-storysource](https://github.com/storybooks/storybook) shows the source code of the template used for the story
- [storybook-addon-vue-info](https://github.com/pocka/storybook-addon-vue-info) detects props info
- [storybook-vue-router](https://github.com/gvaldambrini/storybook-router) add routes to your stories

### Vue Styleguidist

[Vue-styleguidist](https://github.com/vue-styleguidist/vue-styleguidist) is a port of [react-styleguidist](https://github.com/styleguidist/react-styleguidist) for vue components.
If you are building a Design System and want to showcase the powers of your components, this would probably be your first choice.

Example: [vue-styleguide](http://rafaelescala.com/vue-styleguide/) or [vue-design-system](https://vueds.com/example/)

It supports all the following vue features:

- Single file components
- Class style components
- Functional component templates

> - Template and script commenting (Events, Methods, Props, Slots)

Reviewer: ```suggestion
- JSDoc and in-template documentation comments (Events, Methods, Props, Slots)
```

- JSX & TSX
- TypeScript & Flow
- Vuex & Vue-i18n

> - externalized sources for script and template

Reviewer: ```suggestion
- external sources for script and template tags
```
Not sure ppl will get this hehe :D

- styled components

> #### Documentation style

Reviewer: We are deff missing this type of section for the other benches.

Author: added

> styleguidist uses JSDoc:

Reviewer: ```suggestion
Styleguidist uses standard JSDoc comments to extract useful meta data about your components:
```

```javascript
props: {
    /**
     * The color for the button example
     */
    colorExtends: {
      type: String,
      default: '#333'
    }
  }
```

#### Pros

> 

Reviewer: I am missing the JSDoc extraction part.
I am also missing the "generates a fully working doc website".

- Displays a live editor to try components
- Meant for design system creation (powers vue-design-system)

> - A strong React community

Reviewer: Even though its nice to have some community, how would they help if you have a Vue issue?

#### Cons

> - Little configurability on output (HTML only)

Reviewer: ```suggestion
- Generated output is currently limited to just HTML, using predefined template.
```

- JSDoc can be verbose
- Scaffolding built in React

### Vuese

> https://github.com/vuese/vuese

Reviewer: Remove this line

> Vuese is a documentation oriented.

Reviewer: Add a link to Vuese like for other benches.

Reviewer: Also we can merge the below 2 sentences into one paragraph.

It sacrifices a bit of user experience for a lot of integrability.

> If you have to generate your doc to integrate it, this is probably the tool that you want.

Reviewer: ```suggestion
This tool is commonly used to generate some sort of raw output, like markdown files, that can be passed down to a dedicated documentation generator, like Vuepress or Docute.
```

> Example: https://buptsteve.github.io/markdown-it-vuese/example/demo.html#props

Reviewer: Move this to example links below

It supports all the following vue features:

- Single file components
- Class style components
- Template and script commenting (Events, Methods, Props, Slots)
- JSX & TSX
- TypeScript & Flow
- Vuex & Vue-i18n

#### Documentation style

> vuese uses its own documentaion format more compact than JSDoc

Reviewer: ```suggestion
vuese uses its own documentation format, more compact than JSDoc:
```

```javascript
props: {
  // The name of the form
  name: {
    type: String
  }
}
```

#### Pros

- Compact documentation
- Vuepress default output
- Configurable docute or markdown outputs
- Entirely made with VueJs

#### Cons

- Static code samples

### Vue Design System

> https://github.com/viljamis/vue-design-system

Reviewer: Move this link below

Vue designs system is a combination of vue-styleguidist and [salesforce-ux/theo](https://github.com/salesforce-ux/theo) that helps teams build a starter kit for their design system. It is highly opinionated. Choose it if you are starting a new project or if you don't want to think about your own naming.

> vueds documentation is a great read, even if you are not going to use it.

Reviewer: Maybe add the acronym above, next to the name in brackets, so ppl know what you are talking about?

## Documentation Generators

Automatic documentation generators scan your files and output a structured format, that is easy to either use directly or be passed to another tool.

> Such tools can scan your files for JSDoc blocks and output either a json file or full blown websites, ready to deploy.

Reviewer: We have some repetition here with the above line. 
Also how is Vuese different than Vuedoc? Doesnt it generate just plain static markdown file too?

## Component workbenches

A workbench is a development environment for UI components. It allows developers to setup interactive examples of how a UI component can be  used. One can setup a live demo of a Vue component by passing different props to it. Perfect to show your designer or for you to test the result of your work

Workbenches are very powerful when working in big teams with lots of components. They allow other developers to discover how components work by tweaking them. Even better, these showcases can turn into static websites.

> ### Storybook 🌟

Reviewer: I like this, but it generates a super weird anchor link. Maybe I should think of a better way. Lets remove it for now.

Author: removed

- [storybook-addon-vue-info](https://github.com/pocka/storybook-addon-vue-info) detects props info
- [storybook-vue-router](https://github.com/gvaldambrini/storybook-router) add routes to your stories

---

> - Honorable mention to [vue-storybook](https://github.com/mattrothenberg/vue-storybook). In single file components, write stories in a custom `<story>` block!

Author: done

### Vue Styleguidist

> 

Reviewer: Why did we drop the features? Those are actually very useful to compare with others?

#### Documentation style

Styleguidist uses standard JSDoc comments to extract useful meta data from components.

```html
<template>
  <form :name="name">
    {{name}}
    <input type="text">
    <button @click="clear">clear</button>
  </form>
</template>

<script>
/**
  * This is a description of the component
  */
export default {
  name: 'MyComponent',
  props: {
    /**
     * The name of the form, up to 8 characters
     */
    name: {
      type: [String, Number],
      required: true,
      validator () {}
    }
  },
  methods: {
    /**
     * Used to manually clear the form
     * @public
     */
    clear () {
      /**
       * Fire when the form is cleared
       * @property The argument is a boolean value representing xxx
       */
      this.$emit('onclear', true)
    }
  }
}
</script>

<docs>
Here you can use markdown to show uses of the above component.

## The form component can be set with a nane

Here give a reason for using the component. The code samples are going to be rendered automatically with the current comment.

'''vue
<MyComponent name='My nice form'>
'''
</docs>
```

#### Pros

> - Detects all props, events, methods and slots without human intervention

Reviewer: ```suggestion
- Detects props, events, methods and slots automatically
```

- Documentation is done inside the components using standard JSDoc

> - Markdown documentation can live inside custom <docs> block of a Single File Component

Reviewer: ```suggestion
- Markdown documentation can live inside custom `<docs>` block of a Single File Component
```

#### Pros

- Displays a live editor to try components even on the static website
- Meant for design system creation and showcase (powers vue-design-system)
- Multiple components examples can be shown in the same markdown page

> 

Reviewer: I want to stress again, please mention it automagically extracts data from the component, like props, slots, events etc. This is a major PRO when it comes to pure documentation generators.

Author: Oh yeah I forgot the main selling point.

#### Cons

> - Generated output is currently limited to just HTML, using a cussopredefined template.

Author: Well, what brought be to styleguidist in the first place is the customizability of the output.
The website will look and feel exactly as you need it to feel. It makes it a great tool for publishing a good interactive documentation.

- JSDoc can be verbose
- Scaffolding built in React

#### Useful Links

- [Github](https://github.com/vue-styleguidist/vue-styleguidist)
- [Example](http://rafaelescala.com/vue-styleguide/)
- [Documentation](https://vue-styleguidist.github.io/)
- [Discord Channel](https://discordapp.com/channels/325477692906536972/538786416092512278)

### Vuese

[Vuese](https://github.com/vuese/vuese) is documentation oriented. It sacrifices interactivity to be more integrable.
If you have to generate your doc to integrate it in a bigger one, this is the workbench that you want. Without sacrificing the Hot Module Reload, it generates a good looking static doc.

#### Documentation style

Vuese uses its own documentation format, more compact than JSDoc.

```javascript

> props: {

Reviewer: Lets expand the example a bit more, like here 
https://vuese.github.io/vuese-explorer/

#### Cons

- Static code samples

> - Unsupported features of VueJs

Reviewer: Maybe list which ones or rephrase like 
`Not fully compliant with all Vue.js specs` ?

Author: Indeed, I added some examples

#### Useful Links

- [Github](https://github.com/vuese/vuese)
- [Example of markdown-it with vuese](https://buptsteve.github.io/markdown-it-vuese/example/demo.html#props)

### Vue Design System

Vue design system is a combination of [vue-styleguidist](#vue-styleguidist) and [salesforce-ux/theo](https://github.com/salesforce-ux/theo). It goes by *vueds* for short. Vueds is a boilerplate for making your own design systems. It is highly opinionated. If you are starting a new project or if you don't want to think about your naming conventions simply clone the project.

[vueds documentation](https://github.com/viljamis/vue-design-system/wiki) is a great read. Even if you are not going to use it directly, you will love to learn about design systems and how to use them.

#### Useful Links

- [Example](https://vueds.com/example/)
- [Official website](https://vueds.com/)
- [Github](https://github.com/viljamis/vue-design-system)

## Documentation Generators

Automatic documentation generators scan your files and output a structured format. Some create markdown, some create html pages.

### Vuedoc

Vuedoc is using JSDoc to build your documentation in a markdown file.

> Since it is using JSDoc it can easily be combined with styleguidist.

Reviewer: ```suggestion
Since it is using JSDoc, it can easily be combined with styleguidist.
```
How? Is this explained anywhere? If no, then we should not confuse ppl.

Author: Let's not, I will remove it

Author: Removed

Vuese uses its own documentation format, more compact than JSDoc.

```javascript
// This is a description of the component
export default {
  name: 'MyComponent',
  props: {
    // The name of the form, up to 8 characters
    name: {
      type: [String, Number],
      required: true,
      validator () {}
    }
  },
  methods: {
    // @vuese
    // Used to manually clear the form
    clear () {
      // Fire when the form is cleared
      // @arg The argument is a boolean value representing xxx
      this.$emit('onclear', true)
    }
  }
}
```

#### Pros

> - Detects all props, events, methods and slots without human intervention

Reviewer: ```suggestion
- Detects props, events, methods and slots automatically
```
